### PR TITLE
John/fix multicall

### PR DESCRIPTION
**Pull request by johngrantuk**
Update to use correct indexing for onchainData. 
Changed to throw errors in cases where multicall fails.
Added some basic tests for onchainData.
Updated StablePool ABI.

~~Also there appears to be an issue with getCostOfSwapInToken. It was previously using EVM scaled numbers but this causes issues in `optimizeSwapAmounts` because it is using human-scaled numbers. `getNativeAssetPriceInToken` has been changed to use human scaled number.~~

### CODE REVIEW

File: src/swapCost/index.ts [hunk 1]
```diff
@@ -55,12 +51,9 @@ export class SwapCostCalculator {
                 tokenAddress
             );
 
-            const tokenDecimals = await this.getTokenDecimals(tokenAddress);
-
             // Coingecko returns price of token in terms of ETH
             // We want the price of 1 ETH in terms of the token base units
-            const ethPerTokenWei = scale(bnum(ethPerToken), 18 - tokenDecimals);
-            const ethPriceInToken = BONE.div(ethPerTokenWei).dp(0);
+            const ethPriceInToken = bnum(1).div(bnum(ethPerToken));
```
Comment (TomAFrench) [inline src/swapCost/index.ts:56]: Can we revert these changes and perform necessary scaling into human values before passing the resulting value into `optimizeSwapAmounts`? With the work in #148, we'd ideally be able to eventually remove usage of `bignumber.js` entirely so I'd like to confine it to particular areas of the codebase if possible.

File: src/poolCaching/onchainData.ts [hunk 2]
```diff
@@ -33,13 +33,6 @@ export async function getOnChainBalances(
     const multiPool = new Multicaller(multiAddress, provider, abis);
 
     subgraphPools.forEach((pool, i) => {
```
Comment (TomAFrench) [inline src/poolCaching/onchainData.ts:35]:
```suggestion
    subgraphPools.forEach((pool) => {
```
We don't need the index here.

File: src/poolCaching/onchainData.ts [hunk 3]
```diff
@@ -94,19 +87,48 @@ export async function getOnChainBalances(
         }
     >;
 
-    Object.entries(pools).forEach(([poolId, onchainData]) => {
+    try {
+        pools = (await multiPool.execute()) as Record<
```
Comment (TomAFrench) [inline src/poolCaching/onchainData.ts:91]: Ideally we'd fix the return value from `multipool.execute` to be compatible with the `Record`. I'm happy to to the type wrangling in a separate PR though.

**Review by TomAFrench — COMMENTED** [review]
Looks good to me other than the changes to the scaling of token costs although we can split these two changes into two PRs to be merged separately.

Looking at the new code to read onchain balances I think we should prioritise the change from storing the pools as an array to as a mapping. Not crucial for us pushing out a new version though.

File: src/swapCost/index.ts (hunk 1 above)
Comment (johngrantuk) [reply src/swapCost/index.ts:56]: Aren't we already using the same bignumber.js functions in the previous version, i.e. bnum and div?
Scaling to human values before `optimizeSwapAmounts` means we need to get the token decimals from somewhere which isn't readily available as things currently stand. Plus the return amount isn't used anywhere else as far as I know so wouldn't it be clearer to return the value in the format we need it?

Comment (TomAFrench) [reply src/swapCost/index.ts:56]: Fair point about the existing implementation using bnum (this is no longer the case in #148 however), what I mean is that I'd like to avoid using human readable values wherever possible as this we currently need to do a ton of scaling (with associated computational + mental load). 

My point is eventually `optimizeSwapAmounts` can use EVM values so I think it would be best if we had the scaling that we need to do now to happen near there so we can remove it once `optimizeSwapAmounts` no longer needs to take human values.

File: src/poolCaching/onchainData.ts (hunk 3 above)
Comment (johngrantuk) [reply src/poolCaching/onchainData.ts:91]: I've left this for a future PR.

**Review by TomAFrench — APPROVED** [review]
LGTM
